### PR TITLE
screenmarker: add ability to put screen markers around widgets

### PR DESCRIPTION
**Pull request by melkypie**
Gives you the ability to put the screen marker around a widget by clicking on the said widget.

Supersedes/Closes #3952

### CODE REVIEW

**Comment by melkypie** [comment]
PR is ready for review, uses menu entries to get the widget rather than iterating through the widget tree

File: runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerPlugin.java [hunk 1]
```diff
@@ -99,13 +103,18 @@
 	private ScreenMarker currentMarker;
 
 	@Getter
+	@Setter
 	private boolean creatingScreenMarker = false;
+	@Getter
+	@Setter
+	private Rectangle selectedWidgetBounds = null;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerPlugin.java:110]: Reset this on shutdown

File: runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerWidgetHighlight.java [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+import net.runelite.api.widgets.Widget;
+import net.runelite.api.widgets.WidgetInfo;
+import net.runelite.api.widgets.WidgetItem;
+import net.runelite.client.ui.overlay.Overlay;
+import net.runelite.client.ui.overlay.OverlayLayer;
+import net.runelite.client.ui.overlay.OverlayPosition;
+import net.runelite.client.ui.overlay.OverlayPriority;
+
+@Slf4j
+class ScreenMarkerWidgetHighlight extends Overlay
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerWidgetHighlight.java:45]: We usually (or always) use `Overlay` in overlay names

File: runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerWidgetHighlight.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+		this.client = client;
+		setPosition(OverlayPosition.DETACHED);
+		setLayer(OverlayLayer.ABOVE_WIDGETS);
+		setPriority(OverlayPriority.HIGH);
+	}
+
+	@Override
+	public Dimension render(Graphics2D graphics)
+	{
+		if (plugin.isCreatingScreenMarker())
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerWidgetHighlight.java:63]: flip this if

File: runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerWidgetHighlight.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+				plugin.setSelectedWidgetBounds(null);
+				return null;
+			}
+
+			MenuEntry menuEntry = menuEntries[last];
+			final int widgetId = menuEntry.getParam1();
+			final int groupId = WidgetInfo.TO_GROUP(widgetId);
+			final int childId = WidgetInfo.TO_CHILD(widgetId);
+
+			Widget widget = client.getWidget(groupId, childId);
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerWidgetHighlight.java:85]: final, and just pass widget.getChild(param0) to drawHighlight directly

File: runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerWidgetHighlight.java [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+			MenuEntry[] menuEntries = client.getMenuEntries();
+
+			int last = menuEntries.length - 1;
+			if (last < 0)
+			{
+				plugin.setSelectedWidgetBounds(null);
+				return null;
+			}
+
+			MenuEntry menuEntry = menuEntries[last];
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerWidgetHighlight.java:80]: final

File: runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerWidgetHighlight.java [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+		{
+			if (client.isMenuOpen())
+			{
+				plugin.setSelectedWidgetBounds(null);
+				return null;
+			}
+
+			MenuEntry[] menuEntries = client.getMenuEntries();
+
+			int last = menuEntries.length - 1;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerWidgetHighlight.java:73]: final

File: runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerWidgetHighlight.java [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+	{
+		if (plugin.isCreatingScreenMarker())
+		{
+			if (client.isMenuOpen())
+			{
+				plugin.setSelectedWidgetBounds(null);
+				return null;
+			}
+
+			MenuEntry[] menuEntries = client.getMenuEntries();
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerWidgetHighlight.java:71]: final

File: runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerPlugin.java [hunk 8]
```diff
@@ -165,6 +175,26 @@ public void setMouseListenerEnabled(boolean enabled)
 		}
 	}
 
+	public void createWidgetMarker()
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerPlugin.java:178]: This is literally same method as startCreation(Point location)

**Comment by melkypie** [comment]
Made the changes, also fixed a bug where creating the default screen markers, the widgets would still get highlighted alongside.

File: runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerMouseListener.java [hunk 9]
```diff
@@ -45,6 +47,15 @@ public MouseEvent mouseClicked(MouseEvent event)
 			return event;
 		}
 
+		if (SwingUtilities.isLeftMouseButton(event))
+		{
+			Rectangle bounds = plugin.getSelectedWidgetBounds();
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerMouseListener.java:52]: final

File: runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerMouseListener.java [hunk 10]
```diff
@@ -45,6 +47,15 @@ public MouseEvent mouseClicked(MouseEvent event)
...
+			if (bounds != null)
+			{
+				plugin.startCreation(new Point(bounds.x, bounds.y), bounds.getSize());
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerMouseListener.java:55]: bounds.getLocation()

File: runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerPlugin.java [hunk 11]
```diff
@@ -99,13 +103,20 @@
 	private ScreenMarker currentMarker;
 
 	@Getter
+	@Setter
 	private boolean creatingScreenMarker = false;
+	@Getter
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerPlugin.java:108]: empty line above

File: runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerPlugin.java [hunk 12]
```diff
@@ -99,13 +103,20 @@
...
+	private boolean drawing = false;
+	@Getter
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerPlugin.java:110]: empty line above

File: runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerWidgetHighlightOverlay.java [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+import net.runelite.api.MenuEntry;
+import net.runelite.api.widgets.Widget;
+import net.runelite.api.widgets.WidgetInfo;
+import net.runelite.api.widgets.WidgetItem;
+import net.runelite.client.ui.overlay.Overlay;
+import net.runelite.client.ui.overlay.OverlayLayer;
+import net.runelite.client.ui.overlay.OverlayPosition;
+import net.runelite.client.ui.overlay.OverlayPriority;
+
+@Slf4j
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerWidgetHighlightOverlay.java:44]: You are not logging anything

**Pull request merged**